packages/react/src/components/cookieConsent/CookieConsentContext.tsx

```diff
@@ -0,0 +1,249 @@
+  main: Description;
+  details: Description;
+};
+
+export type Category = Description & {
+  checkboxAriaDescription?: string;
+  groups: CookieGroup[];
+};
+
+export type SupportedLanguage = 'fi' | 'sv' | 'en';
+
+export type Content = {
+  requiredCookies?: Category;
+  optionalCookies?: Category;
+  texts: {
+    sections: SectionTexts;
+    ui: UiTexts;
+    tableHeadings: TableData;
+  };
+  language: {
+    languageOptions: { code: SupportedLanguage; label: string }[];
+    current: SupportedLanguage;
+    languageSelectorAriaLabel: string;
+    onLanguageChange: (newLanguage: string) => void;
+  };
+  onAllConsentsGiven?: (consents: ConsentObject) => void;
+  onConsentsParsed?: (consents: ConsentObject, hasUserHandledAllConsents: boolean) => void;
+};
+
+export type CookieConsentContextType = {
+  hasUserHandledAllConsents: () => boolean;
+  content: Content;
+  onAction: CookieConsentActionListener;
+  countApprovedOptional: () => number;
+  areGroupConsentsApproved: (consents: CookieData[]) => boolean;
+};
+
+type CookieConsentContextProps = {
+  cookieDomain?: string;
+  children: React.ReactNode | React.ReactNode[] | null;
+  contentSource: ContentSource;
+};
+
+export const CookieConsentContext = createContext<CookieConsentContextType>({
+  hasUserHandledAllConsents: () => false,
+  content: {} as Content,
+  onAction: () => undefined,
+  countApprovedOptional: () => 0,
+  areGroupConsentsApproved: () => false,
+});
+
+export const getConsentsFromCookieGroup = (groups: CookieGroup[]): ConsentList => {
+  return groups.reduce((ids, currentGroup) => {
+    currentGroup.cookies.forEach((consentData) => {
+      ids.push(consentData.id);
+    });
+    return ids;
+  }, []);
+};
+
+export const Provider = ({ cookieDomain, children, contentSource }: CookieConsentContextProps): React.ReactElement => {
+  const language = contentSource.currentLanguage;
+  const content = useMemo(() => {
+    return createContent(contentSource);
+  }, [language]);
+  const requiredConsents = content.requiredCookies
+    ? getConsentsFromCookieGroup(content.requiredCookies.groups)
+    : undefined;
+  const optionalConsents = content.optionalCookies
+    ? getConsentsFromCookieGroup(content.optionalCookies.groups)
+    : undefined;
+  const consentController = useMemo(
+    () => createConsentController({ requiredConsents, optionalConsents, cookieDomain }),
+    [],
+  );
+  const hasUserHandledAllConsents = () =>
+    consentController.getRequiredWithoutConsent().length === 0 && consentController.getUnhandledConsents().length === 0;
+
+  const mergeConsents = () => ({
+    ...consentController.getRequired(),
+    ...consentController.getOptional(),
+  });
+
+  const [, forceUpdate] = useState<number>(0);
+  const reRender = () => {
+    forceUpdate((p) => p + 1);
+  };
+
+  const save = () => {
+    consentController.save();
+  };
+  const notifyOnAllConsentsGiven = () => {
+    if (content.onAllConsentsGiven && hasUserHandledAllConsents()) {
```

Comment:
Can we sometimes call this so that it does not go inside this if? And if we do, how do we handle that situation?

Reply:
Not at the moment. "onAllConsentsGiven"-callback should never be called without all consents and it is easier to check it here rather than everywhere the `notifyOnAllConsentsGiven` is called.